pkgs/_macro_host/lib/macro_host.dart

```diff
@@ -2,26 +2,91 @@
 // for details. All rights reserved. Use of this source code is governed by a
 // BSD-style license that can be found in the LICENSE file.
 
+import 'dart:io';
+
 import 'package:_macro_builder/macro_builder.dart';
 import 'package:_macro_runner/macro_runner.dart';
 import 'package:_macro_server/macro_server.dart';
+import 'package:dart_model/dart_model.dart';
 import 'package:macro_service/macro_service.dart';
 
-class MacroHost {
+/// Hosts macros: builds them, runs them, serves the macro service.
+///
+/// Tools that want to support macros, such as the Analyzer and the CFE, can
+/// do so by running a `MacroHost` and providing their own `MacroService`.
+class MacroHost implements MacroService {
   final MacroServer macroServer;
+  final ListOfServices services;
   final MacroBuilder macroBuilder = MacroBuilder();
   final MacroRunner macroRunner = MacroRunner();
-  late final HostEndpoint hostEndpoint;
 
-  MacroHost._(this.macroServer, this.hostEndpoint);
+  // TODO(davidmorgan): this should be per macro, as part of tracking per-macro
+  // lifecycle state.
+  Set<int>? _macroPhases;
+
+  MacroHost._(this.macroServer, this.services) {
+    services.services.insert(0, this);
+  }
+
+  /// Starts a macro host serving the provided [service].
+  ///
+  /// The service passed in should handle introspection RPCs, it does not need
+  /// to handle others.
+  ///
+  /// TODO(davidmorgan): make this split clearer, it should be in the protocol
+  /// definition somewhere which requests the host handles.
+  static Future<MacroHost> serve({required MacroService service}) async {
+    final listOfServices = ListOfServices();
+    listOfServices.services.add(service);
+    final server = await MacroServer.serve(service: listOfServices);
+    return MacroHost._(server, listOfServices);
+  }
+
+  /// Whether [name] is a macro according to that package's `pubspec.yaml`.
+  bool isMacro(File packageConfig, QualifiedName name) {
+    // TODO(language/3728): this is a placeholder, use package config when
+    // available.
+    return true;
+  }
 
-  static Future<MacroHost> serve({required MacroHostService service}) async {
-    final server = MacroServer(service: service);
-    final endpoint = await server.serve();
-    return MacroHost._(server, endpoint);
+  /// Determines which phases the macro implemented at [name] runs in.
+  Future<Set<int>> queryMacroPhases(
+      File packageConfig, QualifiedName name) async {
+    if (_macroPhases != null) return _macroPhases!;
+    final macroBundle = await macroBuilder.build(packageConfig, [name]);
+    macroRunner.run(macroBundle: macroBundle, endpoint: macroServer.endpoint);
+    // TODO(davidmorgan): wait explicitly for the MacroStartedRequest to
+    // arrive, remove this hard-coded wait.
+    await Future<void>.delayed(const Duration(seconds: 2));
+    return _macroPhases!;
```

Comment:
The implicit initialization of this field seems pretty error prone. It isn't clear how/what is initializing it. I would see if instead of making it an implicit thing, you could make it an explicit return values of whichever thing initializes it, and assign it that way.